src/utils/UWebSocketTransport.ts

```diff
@@ -35,80 +43,159 @@ export async function createUWebSocketTransport<T extends { isConnected: boolean
+          })
+          currentMessage.future.reject(new Error('Message dropped after max retries'))
           messageQueue.shift()
-        } catch (error: any) {
-          item.future.reject(new Error(`Failed to send message: ${error.message}`))
-          messageQueue.shift()
-          cleanup()
-          return
+          continue
+        }
+
+        const result = processNextMessage(currentMessage)
+
+        if (result === UWebSocketSendResult.SUCCESS || result === UWebSocketSendResult.ERROR) {
+          continue
         }
-        await new Promise((resolve) => setTimeout(resolve, 0))
+
+        // Calculate exponential backoff delay
+        const backoffDelay = Math.min(retryDelayMs * Math.pow(2, currentMessage.attempts), maxBackoffDelayMs)
+
+        // Schedule retry with exponential backoff
+        processingTimeout = setTimeout(() => {
+          processingTimeout = null
+          void processQueue()
+        }, backoffDelay)
+
+        // Exit the loop after scheduling retry
+        return
       }
     } finally {
       isProcessing = false
+    }
+  }
 
-      if (messageQueue.length > 0) {
-        queueProcessingTimeout = setTimeout(() => {
-          void processQueue()
-        }, 1000)
+  function processNextMessage(item: QueuedMessage): UWebSocketSendResult {
+    try {
+      const result = socket.send(item.message, true)
+
+      // Only record metrics for known results
+      if (
+        result === UWebSocketSendResult.SUCCESS ||
+        result === UWebSocketSendResult.BACKPRESSURE ||
+        result === UWebSocketSendResult.DROPPED
+      ) {
+        metrics.observe(
+          'ws_message_size_bytes',
+          { result: UWebSocketSendResult[result].toLowerCase() },
+          item.message.byteLength
+        )
+      }
+
+      logger.debug('Processing queued message', {
+        transportId,
+        result,
+        queueLength: messageQueue.length,
+        messageAttempts: item.attempts,
+        messageSize: item.message.byteLength
+      })
+
+      switch (result) {
+        case UWebSocketSendResult.SUCCESS:
+          item.future.resolve()
+          messageQueue.shift()
+          break
+
+        case UWebSocketSendResult.BACKPRESSURE:
+          metrics.increment('ws_backpressure_events', { result: 'backpressure' })
+          item.attempts++ // Increment attempts to enable exponential backoff
+          break
+
+        case UWebSocketSendResult.DROPPED:
+          // Message was dropped by the underlying library due to maxBackpressure limit.
+          // Do not retry this specific message, reject its future and remove from queue.
+          logger.warn('Message dropped due to backpressure limit', {
+            transportId,
+            messageSize: item.message.byteLength
+          })
+          metrics.increment('ws_backpressure_events', { result: 'dropped' })
+          item.future.reject(new Error('Message dropped due to backpressure limit'))
```

Comment:
The dropped messages should be either rejected or the user disconnected if we've reached a maximum amount of retries. We might want to go with the disconnection logic, as if we drop this message and newer messages eventually arrive due to the backpressure being freed, the client might be kept waiting for a response that could never arrive, right?

Reply:
Is there anything the client can do if it receives this error?